### PR TITLE
gh-121445: C implementation for `fnmatch`

### PR DESCRIPTION
**Pull request by picnixz**
- [x] Update `configure` script.
- [x] C implementation
	- [x] use of `lru_cache`
	- [x] use of `os.path.normcase`
	- [x] `fnmatch.fnmatch`
	- [x] `fnmatch.fnmathcase`
	- [x] `fnmatch.translate`
- [x] Make it work on Windows CI/CD.
- [x] Add documentation
	- [x] C documentation
	- [x] Python documentation
	- [x] NEWS entry
- [x] Increase test coverage

* Issue: gh-121445

### CODE REVIEW

File: Modules/Setup [hunk 1]
```diff
@@ -137,6 +137,7 @@ PYTHONPATH=$(COREPYTHONPATH)
 #_datetime _datetimemodule.c
 #_decimal _decimal/_decimal.c
 #_heapq _heapqmodule.c
+_fnmatch _fnmatchmodule.c
```
Comment (picnixz) [inline Modules/Setup]: I don't really know if I need to do more here... There is no tutorial on how to add a new C extension to the core library so if someone can help me for that :(

Comment (picnixz) [reply Modules/Setup:140]: (And that's the reason why there are failures... I think I committed thr wrong file)

**Comment by picnixz** [comment]
I forgot that MSVC didn't have an implementation of fnmatch... I'll add a guard for Windows so that it uses the pure Python implementation.

File: Modules/_fnmatchmodule.c [hunk 2]
```diff
@@ -0,0 +1,246 @@
...
+    return 1;
+}
+
+static inline int
+posix_fnmatch_encoded(const char *pattern, PyObject *name)
+{
+    if (!validate_encoded_object(name)) {
+        return -1;
+    }
+    // case-insensitive match
```
Comment (picnixz) [inline Modules/_fnmatchmodule.c]: Ok, this one was incorrect. It should have been case-sensititive (I messed up the correct function, my bad. I'll push a revision)

**Comment by picnixz** [comment]
So I'm currently working on making `fnmatch` entirely in C. The translation function would only be used if `fnmatch` is not present. Note that the current implementation has some issues that I have not yet pushed so I'll do it on another day.

(It's also a nice exercice for me in order to familiarize myself with some parts that I did not yet touched)

**Comment by picnixz** [comment]
So, I managed to re-implement the translation algorithm in C but I lost a lot of speed and there are still parts that are buggy. I'll try to fix them tomorrow.

I'm pretty sure I can optimize the translation function even better by interning strings and by avoiding passing twice on the pattern as in the Python implementation. I'm also not sure if the UnicodeWriter is actually that efficient and if I could just simplify my life using lists and `join()` at the end.

**Comment by picnixz** [comment]
Ok I think I know why the build fails. It's because I did not regenerate the `configure` script... But my version of autoconf is different from the required one and this makes a lot of changes to other files... so I'll just try to make it work with docker?

**Comment by picnixz** [comment]
Ok, so now it remains to see how I can compile it on Windows... I'll also need to fix an issue where `fnmatch(3)` doesn't know whether a backslash is meant to be escaped or not... I'll need to think of that issue. I think we can still gain speed just for the translation unit.

**Comment by picnixz** [comment]
So, everything should be ready. Documentation and NEWS entry need to be written and I also need to know how I can make it work on Windows.

@zooba I know you are well-versed on Windows platforms so could you help me there please if you have time? How can I specify the files to be compiled correctly...?

**Comment by zooba** [comment]
Easiest way is to look at any other similar module, find all the references to it, and set yours up the same way. Looks like you'll have a bit of rearranging to do here (e.g. header goes somewhere under `Include`).

Filename matching should also be a subset of regex, so it may be easier and more efficient to write our own comparison function. That would also then be fully portable, as well as being easier to handle compatibility.

**Comment by picnixz** [comment]
> Easiest way is to look at any other similar module, find all the references to it, and set yours up the same way. Looks like you'll have a bit of rearranging to do here (e.g. header goes somewhere under Include).

Thank you for that (actually, that's how I managed to do it for the linux build...). I actually never used MSVC so I don't know how to compile nor what I should actually specify but I'll do it the old way (trial and error).

> Filename matching should also be a subset of regex, so it may be easier and more efficient to write our own comparison function. That would also then be fully portable, as well as being easier to handle compatibility.

For now, I've just replicated the Python implementation which translates UNIX shell patterns to RE patterns. Strictly speaking, I don't think the translation is needed when comparing filenames (since we would anyway use a subset of regex indeed), but I didn't want to introduce a new algorithm (which I more or less gave up to describe after 10 minutes...)

**Comment by picnixz** [comment]
Mmmh... so the improvements are.... not improvements =/ Well I'll just forget about using interned strings actually.

**Comment by picnixz** [comment]
I think I'll leave the implementation like that for now and open it for review. At least, we have 1.7x performance gain for `fnmatch.translate` (but let's say we only have 1.5x to be very conservative).

**Comment by picnixz** [comment]
err... why is there this whitespace change?

**Pull request closed without merging**